src/v/cloud_storage/remote_segment.cc

```diff
@@ -499,10 +496,22 @@ remote_segment_batch_reader::read_some(
 
         _cur_ot_state = ot_state;
         auto deferred = ss::defer([this] { _cur_ot_state = std::nullopt; });
-        auto bytes_consumed = co_await _parser->consume();
-        if (!bytes_consumed) {
-            co_return bytes_consumed.error();
+        auto new_bytes_consumed = co_await _parser->consume();
+        if (!new_bytes_consumed) {
+            co_return new_bytes_consumed.error();
+        }
+        if (
+          _bytes_consumed != 0 && _bytes_consumed == new_bytes_consumed.value()
+          && !_config.over_budget) {
+            throw std::runtime_error(fmt::format(
```

Comment:
when throwing exception it is usually a good idea to use `fmt_with_ctx` instead of `fmt::format`. The former will add line numbers etc.. to the error message which can be useful because where the exception is caught and logged is often far away from where it was thrown.